### ISSUE TITLE
Where in the World is Carmen Sandiego (1986) locks up when pressing F12 to change to composite

### ISSUE THREAD

**Issue by Jkapp76** [dosbox-staging/dosbox-staging#4051]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

Latest official

### What Operating System are you using?

Windows 11

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

When loading the game "Where in the World is Carmen Sandiego 1986" Once the police officer begins to walk across the screen Tap F12 on the CGA machine setting to change into composite video. This will change to composite but lock the game.
I only wrote this as a bug because F12 does nothing on Dosbox-x. If you start the game in composite mode it works fine with more colors.
Same with PCJr and Tandy. This is the only composite game I've seen that locks when changing this.

I hope this isn't too small of a nitpick for a bug report.

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Load the game in CGA machine mode.
2. Wait for graphics to appear and press F12.
3. It locks and can't be unlocked.

### Your configuration

```ini
CGA
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
@Jkapp76 Could you check earlier Staging versions and PCem/86box?

**Comment by Jkapp76** [comment]
Okay, staging 0.81 works the same and locks when pressing F12.
86Box works perfect and does nothing at all when tapping F12 just like dosbox-x.

**Comment by interloper98** [comment]
I can't reproduce it; tried `main` and 0.82, as well as `output = texture` and default w/ opengl.

I tried grabbing a recording, but the AVIs get split after each video mode change, so I screencap'd after F12 switches:

![image0023-rendered](https://github.com/user-attachments/assets/d56815fc-0960-4e75-a3b2-2dc67831b8c9)

![image0024-rendered](https://github.com/user-attachments/assets/e145f89c-a008-4341-8412-7f9bf53ea4bc)

![image0025-rendered](https://github.com/user-attachments/assets/458354b9-5cf2-491c-bd2e-78791ce9477f)

![image0026-rendered](https://github.com/user-attachments/assets/32d0d480-fa76-43ce-ab55-8593b3cf22af)

@Jkapp76, here's my directory; maybe my conf or files differ? 

[where_in_the_world_is_carmen_sandiego.zip](https://github.com/user-attachments/files/17693296/where_in_the_world_is_carmen_sandiego.zip)

I also noticed my F12 wasn't assigned to the composite action, so I used the Ctrl+F1 mapper to assign it.

**Comment by Jkapp76** [comment]
I tried your copy of the game that was in your zip, and I changed my output to texture like yours.

Mine still locks when I press F12. I don't know why, but if it's only my PC then it's probably not a bug. Hitting ESC does get me to the next game screen where I can type my name, so this only happens during the intro animation for me.

**Comment by johnnovak** [comment]
> I tried your copy of the game that was in your zip, and I changed my output to texture like yours.
> 
> Mine still locks when I press F12. I don't know why, but if it's only my PC then it's probably not a bug. Hitting Space does get me to the next game screen where I can type my name, but the animation is skipped. So this only happens during the intro animation for me. And the game will lock up again if I switch to or from composite again... which can be resumed again with the space bar.
> 
> Seems weird to me, but maybe it's me that's weird here.

Quite mysterious, indeed! 😆

**Comment by interloper98** [comment]
Darn; I first tested on the Raspberry Pi (where I can switch back and fourth), but now on Windows 11 I'm getting the same "soft ock" behavior as @Jkapp76. 

DOSBox Staging itself is still running, the game is still running, but once composite is enabled the game doesn't update the next screen until (any) key is pressed.

**Comment by johnnovak** [comment]
> DOSBox Staging itself is still running, the game is still running, but once the video-mode change happens the game doesn't update the next screen until (any) key is pressed.

Try with `glshader = sharp` to rule out the auto CRT shader switching thing. That has some platform-specific quirks; I had to massage it a bit to get it working on all three platforms after my initial macOS implementation.

**Comment by Jkapp76** [comment]
I tested with glshader=sharp. Same issue.

**Comment by interloper98** [comment]
There's something about the bare F12 keystroke that's interacting with the composite change at the same time in the game.

I suspect when DOSBox-X and 86Box switch to composite, they're not sending in an F12 keystroke to the game at the same time.

(in DOSBox Staging XT machine types, it looks like the F12 key is doing double-duty: sending the F12 keystroke into the game *and* making the video mode switch to composite. Whereas other hotkeys like Ctrl+F12, Ctrl+F7, etc have modifiersthat hide those F-keys from also hitting the game).

If you fully unmap F12 (Ctrl+F1 mapper, click mouse on F12 key, click "remove bind", then "save") and then exclusively mapping the composite change to F12 (mouse click on the 'CGA Comp' box, then "remove bind", then "add bind", then type an F12 keystroke, then "save") now you'll have the F12 key *only* doing the composite switch and the game won't be aware that F12 was pressed.

This works!

https://github.com/user-attachments/assets/b4f561aa-c195-475e-9af8-456d0983f30c

**Comment by Jkapp76** [comment]
This works!
I also hit F11 to change the composite to NEW composite and this action also freezes the game. I'm having issues figuring out how to do this for F11, but I think this method is the answer.
It would be nice if we could save it this way... unless I just haven't figured that out yet.

**Comment by johnnovak** [comment]
> I suspect when DOSBox-X and 86Box switch to composite, they're not sending in an F12 keystroke to the game at the same time.

That is the correct way, yeah. Once some keystroke combo triggers an action, that keypress combo is "consumed"; we should never send it to the emulated environment _as well_.

> (in DOSBox Staging XT machine types, it looks like the F12 key is doing double-duty: sending the F12 keystroke into the game _and_ making the video mode switch to composite. Whereas other hotkeys like Ctrl+F12, Ctrl+F7, etc have modifiers that hide those F-keys from also hitting the game).

Only in the XT type? In any case, that's just wrong.

Dunno how hard this would be, but we should generically consume any hotkey presses and not forward it to the emulated DOS environment as well. It's either / or.

**Comment by Jkapp76** [comment]
It seems to me that you can use F12 as a hotkey and it works, and you can remove the hotkey and use F12 as a normal F12 key and it also works. But somehow the key is doing both as the default option and the two functions together cause it to hang.

**Comment by interloper98** [comment]
Once some keystroke combo triggers an action, that keypress combo is "consumed"; we should never send it to the emulated environment as well.

Yup; I think that's normally the case in DOSBox because the combo involve ctrl/alt modifiers; the problem is this F12 composite action has not modifiers.. it's just straight up F12.

I suspect the fix here is to make the composite events have modifiers too.

DOSBox's mapper allows multiple event bindings. For example, you could bind the up arrow button to take screenshots, but it by default is also bound to sending in the up-arrow keystroke into DOS (both would happen on button press).


> Only in the XT type? In any case, that's just wrong.

When machine is pcjr/cga/tandy, it looks like the composite stuff lights up with the F12 and F11 extra bindings as well as the CGA composite event box being available in the mapper to bind:

![xt-mapper](https://github.com/user-attachments/assets/ab826cdb-d3a6-4232-8232-b61de5b77a85)

@Jkapp76, as for the composite new/old toggle (that you mentioned F11 is doing), I'm not seeing it available in the mapper, so I'm not sure if it can be re-bound using the mapper. 

> Dunno how hard this would be, but we should generically consume any hotkey presses and not forward it to the emulated DOS environment as well. It's either / or.

From what it looks like, the mapper allows physical host events (keyboard/mouse/joystick) to be bound to multiple logical events. The fix would be to bump out the prior binding and only allow one to exist (just guessing).

**Comment by johnnovak** [comment]
> DOSBox's mapper allows multiple event bindings. For example, you could bind the up arrow button to take screenshots, but it by default is also bound to sending in the up-arrow keystroke into DOS (both would happen on button press).

Yes, so my view is that's wrong. But I'm not saying to rewrite the mapper now; I think this is a minor issue and people can live with it until the mapper rewrite.

> From what it looks like, the mapper allows physical host events (keyboard/mouse/joystick) to be bound to multiple logical events. The fix would be to bump out the prior binding and only allow one to exist (just guessing).

Yeah, if it's easy. I don't think mapping multiple logical events to a single physical event/combo is desirable or very useful. Sure, one can always come up with artificual or 0.1% examples, but I'd say in the new mapper there will always be a 1:N mapping between logical and physical events (so you can toggle the same logical event by multiple physical events, but you can never trigger multiple logical events with the same physical event or event combo).

Then, if it's an easy and low risk change, yeah we could do this in the existing mapper too, but I wouldn't spend too much time on it.

**Comment by Jkapp76** [comment]
@interloper98 
The F11 key changes New/Old composite mode and is labeled "INC Knob" on the keyboard.

I used your map/unmap method and can fix both keys that way. This game has better colors with the New setting so I prefer to tap that once for this game.

I recently created a compilation of all Carmen Sandiego games for my Nephews in one dosbox installer. Their only complaint was they wish Dosbox Staging had a Red X in the upper right corner to easily shut off the emulation when they're done... I thought about requesting this as an option, but figured I'd get a few laughing emojis for it.

**Comment by Jkapp76** [comment]
I just tested Where in the "USA" is Carmen Sandiego to see if this issue happens on that USA version like it does on the World version and it does. So here's at least two games that do this and I'd say they're both popular games... although few people probably know to use composite video to get the extra colors.

**Comment by johnnovak** [comment]
Reopened because of the above revert.
